lightning/src/routing/router.rs

```diff
@@ -1520,12 +1529,91 @@ where L::Target: Logger {
+			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+
+			// Init PRNG with path nonce
+			let mut path_nonce = [0u8; 8];
+			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..8]);
+			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+			let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
+
+			// Pick a random path length in [1 .. 3]
+			prng.process_in_place(&mut random_path_bytes);
+			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
+
+			for _random_hop in 0..random_walk_length { 
+				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+					// Randomly choose the next hop
+					prng.process_in_place(&mut random_path_bytes);
+					if let Some(random_index) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len()) {
+						if let Some(random_channel_id) = cur_node.channels.get(random_index) {
+							if let Some(random_channel) = network_channels.get(random_channel_id) {
+								if random_channel.node_one == cur_node_id {
+									cur_node_id = random_channel.node_two;
+									if let Some(dir_info) = &random_channel.one_to_two {
+										shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+											.checked_add(dir_info.cltv_expiry_delta.into()).unwrap_or(shadow_ctlv_expiry_delta_offset);
+									}
+								} else {
+									cur_node_id = random_channel.node_one;
+									if let Some(dir_info) = &random_channel.two_to_one {
+										shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+											.checked_add(dir_info.cltv_expiry_delta.into()).unwrap_or(shadow_ctlv_expiry_delta_offset);
+									}
+								};
+							}
+						}
+					}
+				}
+			}
+		} else {
+			// If the entire path is private, choose a random offset from multiples of 40, i.e., the most used cltv_expiry_delta
+			let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 8]);
+			let mut random_bytes = [0u8; 4];
+			prng.process_in_place(&mut random_bytes);
+			let random_walk_length = u32::from_be_bytes(random_bytes).wrapping_rem(3).wrapping_add(1);
+			shadow_ctlv_expiry_delta_offset = random_walk_length.wrapping_mul(40);
+		}
+
+		// Limit the total offset to reduce the worst-case locked liquidity timevalue
+		const MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET: u32 = 432; // 3*144
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET);
+
+		// Limit the offset so we never exceed the max_total_cltv_expiry_delta
+		let max_path_offset = payment_params.max_total_cltv_expiry_delta
+			.checked_sub(path.iter().fold(0, |acc,b| b.cltv_expiry_delta.max(acc)))
+			.unwrap_or(shadow_ctlv_expiry_delta_offset);
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, 
+												   max_path_offset.checked_sub(max_path_offset.wrapping_rem(40)).unwrap_or(max_path_offset));
+
+		// Add 'shadow' CLTV offset to all hops but the final one
```

Comment:
After thinking about it a bit more, I'm starting to reconsider whether adding the offset to all but the final hop might be the wrong approach? 
As https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L48 describes the `RouteHop`s returned from `get_route` hold the `cltv_expiry_delta` for that specific hop that will be added/aggregated later? Therefore adding the offset for the entire route once to the first hop should suffice here?

Reply:
Uh, yes, I think you're correct, we should add the CLTV delta only on the last hop (I assume - we want the final recipient to get the extra CLTV, no, not the hop-before-last?)

Reply:
Ah, right! I was thinking that adding an offset to the last one would be superfluous since the payee knows anyhow that it's the final hop. However, it's of course correct that the last hop is the one *to* the payee, i.e., the one that the second-to-last node can still observe. Fixed with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/67137c6fb5cf803dcd7031dab6eb0c39bd9fdbca.